param_sync.py

```diff
@@ -7,12 +7,39 @@ class ParamSyncRule:
     training.
     """
 
+    def update_core(self, local_params, master_params):
+        """
+        Iterator over updates for local and master params.
+
+        This function must be ready to deal with symbolic or numerical
+        inputs.
+        """
+        raise NotImplementedError()
+
     def update_params(self, local_params, master_params):
         """
         Perform an inplace update of the master parameters and returns the
         parameters based on a certain parameter synchronisation rule.
         """
-        raise NotImplementedError()
+        for lp, mp, ul, um in izip(local_params, master_params,
+                                   self.update_core(local_params,
+                                                    master_params)):
+            lp += ul
+            mp += um
+
+    def update_fn(self, local_params):
+        """
+        Return a theano function that will update the local shared
+        params and return the new values for the master params.
+        """
+        from theano import function
```

Comment:
This seem a good reason for me. But I was under the impression that Theano
was requested by platoon. You mean that with this import there, Theano is
optional?

If so, we should at least add a comment why we break good practice.

On Wed, Jan 6, 2016 at 5:20 PM, abergeron <notifications@github.com> wrote:

> In param_sync.py
> <https://github.com/mila-udem/platoon/pull/11#discussion_r49017898>:
>
> >      def update_params(self, local_params, master_params):
> >          """
> >          Perform an inplace update of the master parameters and returns the
> >          parameters based on a certain parameter synchronisation rule.
> >          """
> > -        raise NotImplementedError()
> > +        for lp, mp, ul, um in izip(local_params, master_params,
> > +                                   self.update_core(local_params,
> > +                                                    master_params)):
> > +            lp += ul
> > +            mp += um
> > +
> > +    def update_fn(self, local_params):
> > +        """
> > +        Return a theano function that will update the local shared
> > +        params and return the new values for the master params.
> > +        """
> > +        from theano import function
>
> "never" is a big word.
>
> Our policy is more like "not without a good reason" and my reason is to
> not force a dependency on theano if not strictly required.
>
> We can debate if that is "good" or not though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/platoon/pull/11/files#r49017898>.
>